C compilation error when compiling simple Generic Interface test case

### Describe the bug

The vlang compiler hits a builder error when compiling a simple test case using generic interfaces.

### Reproduction Steps

Test case:

```
❯ cat src/main.v
module main

interface Foo[T] {
        val T
        foo() T
}

struct Bar[T] implements Foo[T] {
        val T
}

fn (b Bar[T]) foo() T {
        return b.val
}

fn main() {
        b := Bar{val: 0}
        b.foo()
}
```
Compilation attempt:

```
❯ ~/Work/repos/vlang/v/v run .
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/generic_interface1.01J6YJCHPVBPEXFB083RKKMAPB.tmp.c:998: error: ';' expected (got "T")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

C output:

```
❯ ~/Work/repos/vlang/v/v -show-c-output run .
======== Output of the C Compiler (/home/robin/Work/repos/vlang/v/thirdparty/tcc/tcc.exe) ========
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:998: error: ';' expected (got "T")
==================================================================================================
======== Output of the C Compiler (cc) ========
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:998:9: error: unknown type name ‘T’
  998 |         T* val;
      |         ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2170:65: error: unknown type name ‘main__Bar_T_T’; did you mean ‘main__Bar_T_int’?
 2170 | static main__Foo_T_T I_main__Bar_T_T_to_Interface_main__Foo_T_T(main__Bar_T_T* x);
      |                                                                 ^~~~~~~~~~~~~
      |                                                                 main__Bar_T_int
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2177:15: error: unknown type name ‘T’
 2177 | static inline T main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper(main__Bar_T_T* b) {
      |               ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2177:78: error: unknown type name ‘main__Bar_T_T’; did you mean ‘main__Bar_T_int’?
 2177 | static inline T main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper(main__Bar_T_T* b) {
      |                                                                              ^~~~~~~~~~~~~
      |                                                                              main__Bar_T_int
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2182:9: error: expected specifier-qualifier-list before ‘T’
 2182 |         T (*_method_foo)(void* _);
      |         ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2187:18: error: ‘struct _main__Foo_T_T_interface_methods’ has no member named ‘_method_foo’
 2187 |                 ._method_foo = (void*) main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper,
      |                  ^~~~~~~~~~~
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2187:40: error: ‘main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper’ undeclared here (not in a function)
 2187 |                 ._method_foo = (void*) main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper,
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2187:32: warning: excess elements in struct initializer
 2187 |                 ._method_foo = (void*) main__Bar_T_T_foo_T_T_Interface_main__Foo_T_T_method_wrapper,
      |                                ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2187:32: note: (near initialization for ‘main__Foo_T_T_name_table[0]’)
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2190:18: error: ‘struct _main__Foo_T_T_interface_methods’ has no member named ‘_method_foo’
 2190 |                 ._method_foo = (void*) 0,
      |                  ^~~~~~~~~~~
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2190:32: warning: excess elements in struct initializer
 2190 |                 ._method_foo = (void*) 0,
      |                                ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2190:32: note: (near initialization for ‘main__Foo_T_T_name_table[1]’)
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2196:72: error: unknown type name ‘main__Bar_T_T’; did you mean ‘main__Bar_T_int’?
 2196 | static inline main__Foo_T_T I_main__Bar_T_T_to_Interface_main__Foo_T_T(main__Bar_T_T* x) {
      |                                                                        ^~~~~~~~~~~~~
      |                                                                        main__Bar_T_int
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c: In function ‘I_voidptr_to_Interface_main__Foo_T_T’:
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2209:25: error: ‘T’ undeclared (first use in this function)
 2209 |                 .val = (T*)((char*)x/*.... ast.voidptr_type */),
      |                         ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2209:25: note: each undeclared identifier is reported only once for each function it appears in
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2209:27: error: expected expression before ‘)’ token
 2209 |                 .val = (T*)((char*)x/*.... ast.voidptr_type */),
      |                           ^
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c: In function ‘v_typeof_interface_main__Foo_T_T’:
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2260:21: error: ‘_main__Foo_T_T_main__Bar_index’ undeclared (first use in this function); did you mean ‘_main__Foo_T_T_main__Bar_T_T_index’?
 2260 |         if (sidx == _main__Foo_T_T_main__Bar_index) return "Bar";
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     _main__Foo_T_T_main__Bar_T_T_index
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c: In function ‘v_typeof_interface_idx_main__Foo_T_T’:
/tmp/v_1000/generic_interface1.01J6YJCXYM7QMQRN0QAVB6XHBA.tmp.c:2266:21: error: ‘_main__Foo_T_T_main__Bar_index’ undeclared (first use in this function); did you mean ‘_main__Foo_T_T_main__Bar_T_T_index’?
 2266 |         if (sidx == _main__Foo_T_T_main__Bar_index) return 102;
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     _main__Foo_T_T_main__Bar_T_T_index
===============================================
No such file or directory; code: 2
```

Versions:

```
❯ ~/Work/repos/vlang/v/v -version
V 0.4.7 26cebc6

```

### Expected Behavior

Successful compilation and run.

### Current Behavior

Builder error. Please see above for details.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

 0.4.7 26cebc6

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.7 515b106.26cebc6
OS: linux, "Fedora Linux 40 (Workstation Edition)"
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics

getwd: /home/robin/Work/spaces/vlang/experiments/generic_interface1
vexe: /home/robin/Work/repos/vlang/v/v
vexe mtime: 2024-09-04 07:57:23

vroot: OK, value: /home/robin/Work/repos/vlang/v
VMODULES: OK, value: /home/robin/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.0
Git vroot status: weekly.2024.34-81-g26cebc6b
.git/config present: true

CC version: cc (GCC) 14.2.1 20240801 (Red Hat 14.2.1-1)
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/table.v
+++ b/vlib/v/ast/table.v
@@ -2087,7 +2087,7 @@ pub fn (mut t Table) unwrap_generic_type(typ Type, generic_names []string, concr
 				}
 			}
 			mut info := ts.info
-			info.is_generic = false
+			info.is_generic = final_concrete_types.any(it.has_flag(.generic))
 			info.concrete_types = final_concrete_types
 			info.parent_type = typ.set_flag(.generic)
 			info.fields = fields

--- a/vlib/v/tests/interface_generic_implements_test.v
+++ b/vlib/v/tests/interface_generic_implements_test.v
@@ -1,0 +1,19 @@
+interface Foo[T] {
+	val T
+	foo() T
+}
+
+struct Bar[T] implements Foo[T] {
+	val T
+}
+
+fn (b Bar[T]) foo() T {
+	return b.val
+}
+
+fn test_main() {
+	b := Bar{
+		val: 0
+	}
+	assert b.foo() == 0
+}

